langchain/src/experimental/masking/parser.ts

```diff
@@ -0,0 +1,139 @@
+import { MaskingTransformer } from "./transformer.js";
+import { MaskingParserConfig } from "./types.js";
+
+/**
+ * MaskingParser class for handling the masking and rehydrating of messages.
+ */
+export class MaskingParser {
+  private transformers: MaskingTransformer[];
+  private state: Map<string, string>;
+  private config: MaskingParserConfig;
+
+  constructor(config: MaskingParserConfig = {}) {
+    this.transformers = config.transformers || [];
+    this.state = new Map<string, string>();
+    this.config = config;
+  }
+
+  /**
+   * Adds a transformer to the parser.
+   * @param transformer - An instance of a class extending MaskingTransformer.
+   */
+  addTransformer(transformer: MaskingTransformer) {
+    this.transformers.push(transformer);
+  }
+
+  /**
+   * Getter method for retrieving the current state.
+   * @returns The current state map.
+   */
+  public getState(): Map<string, string> {
+    return this.state;
+  }
+
+  /**
+   * Masks the provided message using the added transformers.
+   * This method sequentially applies each transformer's masking logic to the message.
+   * It utilizes a state map to track original values corresponding to their masked versions.
+   *
+   * @param message - The message to be masked.
+   * @returns A masked version of the message.
+   * @throws {TypeError} If the message is not a string.
+   * @throws {Error} If no transformers are added.
+   */
+  async parse(message: string): Promise<string> {
+    this.config.onMaskingStart?.(message);
+
+    // Check if there are any transformers added to the parser. If not, throw an error
+    // as masking requires at least one transformer to apply its logic.
+    if (this.transformers.length === 0) {
+      throw new Error(
+        "MaskingParser.parse Error: No transformers have been added. Please add at least one transformer before parsing."
+      );
+    }
+
+    if (typeof message !== "string") {
+      throw new TypeError(
+        "MaskingParser.parse Error: The 'message' argument must be a string."
+      );
+    }
+
+    // Initialize the variable to hold the progressively masked message.
+    // It starts as the original message and gets transformed by each transformer.
+    let processedMessage = message;
+
+    // Iterate through each transformer added to the parser.
+    this.transformers.forEach((transformer) => {
+      // Apply the transformer's transform method to the current state of the message.
+      // The transform method returns a tuple containing the updated message and state.
+      // The state is a map that tracks the original values of masked content.
+      // This state is essential for the rehydration process to restore the original message.
+      [processedMessage, this.state] = transformer.transform(
```

Comment:
I think it might be nice to allow `async` transform functions instead of just sync?

Reply:
Only other comment here is that I don't love mutating `this.state` internally within `transformer.transform` - could see some nasty bugs where a transformer implementation e.g. mistakenly clears state from previous transformers.

My preference would be to manually combine state at the parser level based on each transformer's output, but fine if you want to stick with this since it's in experimental.